File: src/parser/parser_template.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_map>
#include "json.hpp"

using json = nlohmann::json;
using namespace std;

int main(int argc, char* argv[]) {
    if (argc < 3) {
        cout << "Usage: parser parse-table.json \"token1 token2 ...\" [override.json]\n";
        return 1;
    }

    string table_file = argv[1];
    string token_string = argv[2];
    string override_file = (argc > 3) ? argv[3] : "";

    /* tokenize input */
    vector<string> tokens;
    string temp;
    for (char c : token_string) {
        if (c == ' ') {
            if (!temp.empty()) {
                tokens.push_back(temp);
                temp.clear();
            }
        } else {
            temp += c;
        }
    }
    if (!temp.empty()) tokens.push_back(temp);
    tokens.push_back("$");

    /* load parse table */
    ifstream f(table_file);
    if (!f.is_open()) {
        cout << "Cannot open " << table_file << endl;
        return 1;
    }

    json data;
    f >> data;

    auto parse_table = data["parseTable"]["_table"];

    unordered_map<string, pair<string, vector<string>>> productions;
    for (auto& p : data["productions"]) {
        string num = to_string((int)p["productionNumber"]);
        string lhs = p["lhs"];
        vector<string> rhs = p["rhs"].get<vector<string>>();
        productions[num] = { lhs, rhs };
    }

    /* load override table if exists */
    json override_table;
    if (!override_file.empty()) {
        ifstream of(override_file);
        if (!of.is_open()) {
            cout << "Cannot open override file\n";
            return 1;
        }
        of >> override_table;
    }

    vector<string> stack;
    stack.push_back("1");
    int index = 0;

    while (true) {
        string state = stack.back();
        string token = tokens[index];

        json entry;

        if (!override_file.empty() &&
            override_table.contains(state) &&
            override_table[state].contains(token)) {
            entry = override_table[state][token];
        } else {
            if (!parse_table.contains(state) ||
                !parse_table[state].contains(token)) {
                cout << "Syntax Error\n";
                cout << "State: " << state
                     << "\nUnexpected token: " << token
                     << "\nPosition: " << index << endl;
                return 1;
            }
            entry = parse_table[state][token];
            if (entry.is_array()) {
                cout << "Error: conflict in state "
                     << state << " with token " << token << endl;
                return 1;
            }
        }

        string action_type = entry["type"];

        if (action_type == "accept") {
            cout << "Input Accepted\n";
            break;
        }

        else if (action_type == "shift") {
            stack.push_back(token);
            stack.push_back(to_string((int)entry["destination"]));
            index++;
        }

        else if (action_type == "reduce") {
            string rule_num = to_string((int)entry["ruleNumber"]);
            auto [lhs, rhs] = productions[rule_num];

            for (size_t i = 0; i < rhs.size() * 2; i++)
                stack.pop_back();

            string top_state = stack.back();
            auto go = parse_table[top_state][lhs];
            stack.push_back(lhs);
            stack.push_back(to_string((int)go["destination"]));
        }

        else if (action_type == "shift_reduce") {
            if (entry.contains("destination")) {
                stack.push_back(token);
                stack.push_back(to_string((int)entry["destination"]));
                index++;
            }
            else if (entry.contains("ruleNumber")) {
                string rule_num = to_string((int)entry["ruleNumber"]);
                auto [lhs, rhs] = productions[rule_num];

                for (size_t i = 0; i < rhs.size() * 2; i++)
                    stack.pop_back();

                string top_state = stack.back();
                auto go = parse_table[top_state][lhs];
                stack.push_back(lhs);
                stack.push_back(to_string((int)go["destination"]));
            }
            else {
                cout << "Error: shift_reduce entry malformed\n";
                return 1;
            }
        }

        else {
            cout << "Unknown action type: " << action_type << endl;
            return 1;
        }
    }

    return 0;
}

File: src/parser/parser_template.py
import json
import sys


def load_json(path):
    with open(path, "r") as f:
        return json.load(f)


def terminate_with_error(state, token, pos):
    print("Syntax Error")
    print(f"State: {state}\nUnexpected token: {token}\nPosition: {pos}")
    sys.exit(1)


def apply_reduce(stack, rule_id, prod_map, table):
    lhs, rhs = prod_map[rule_id]

    pop_count = len(rhs) * 2
    for _ in range(pop_count):
        stack.pop()

    current_state = str(stack[-1])
    goto_entry = table[current_state][lhs]

    stack.extend([lhs, goto_entry["destination"]])


def main():
    if len(sys.argv) < 3:
        print('Usage: python parser.py parse-table.json "token1 token2 ..." [override.json]')
        sys.exit(1)

    table_path = sys.argv[1]
    input_tokens = sys.argv[2].split()
    override_path = sys.argv[3] if len(sys.argv) > 3 else None

    input_tokens.append("$")

    grammar_data = load_json(table_path)

    table = grammar_data["parseTable"]["_table"]
    production_rules = {
        str(p["productionNumber"]): (p["lhs"], p["rhs"])
        for p in grammar_data["productions"]
    }

    overrides = load_json(override_path) if override_path else {}

    parse_stack = [1]
    cursor = 0

    while True:
        current_state = str(parse_stack[-1])
        current_token = input_tokens[cursor]

        if current_state in overrides and current_token in overrides[current_state]:
            action = overrides[current_state][current_token]
        else:
            if current_state not in table or current_token not in table[current_state]:
                terminate_with_error(current_state, current_token, cursor)

            action = table[current_state][current_token]

            if isinstance(action, list):
                print(f"Error: conflict in state {current_state} with token {current_token}")
                sys.exit(1)

        kind = action["type"]

        if kind == "accept":
            print("Input Accepted")
            break

        if kind == "shift":
            parse_stack.extend([current_token, action["destination"]])
            cursor += 1
            continue

        if kind == "reduce":
            rule_key = str(action["ruleNumber"])
            apply_reduce(parse_stack, rule_key, production_rules, table)
            continue

        if kind == "shift_reduce":
            if "destination" in action:
                parse_stack.extend([current_token, action["destination"]])
                cursor += 1
            elif "ruleNumber" in action:
                rule_key = str(action["ruleNumber"])
                apply_reduce(parse_stack, rule_key, production_rules, table)
            else:
                print("Error: shift_reduce entry malformed")
                sys.exit(1)
            continue

        print(f"Unknown action type: {kind}")
        sys.exit(1)


if __name__ == "__main__":
    main()
